File: src/availability-processor/golang/src/handler.go
package processor

import (
	"github.com/birdiecare/availability-processor-exercise/src/repositories"
)

// EventProcessor handles caregiver availability events
type EventProcessor struct {
	visitRepo repositories.VisitRepository
}

// NewEventProcessor creates a new event processor
func NewEventProcessor(visitRepo repositories.VisitRepository) *EventProcessor {
	return &EventProcessor{
		visitRepo: visitRepo,
	}
}

func (p *EventProcessor) HandleEvent(event CaregiverPermanentUnavailabilityEvent) error {
	futureDate := event.EffectiveFrom.AddDate(1, 0, 0) // 1 year in the future

	visits, err := p.visitRepo.GetCalendar(
		&event.CaregiverId,
		event.EffectiveFrom,
		futureDate,
	)
	if err != nil {
		return err
	}

	// Unassign all visits that occur after the permanent unavailability starts
	for _, visit := range visits {
		if visit.StartTime.After(event.EffectiveFrom) || visit.StartTime.Equal(event.EffectiveFrom) {
			err := p.visitRepo.Unassign(visit.Id, event.CaregiverId)
			if err != nil {
				return err
			}
		}
	}

	return nil
}

File: src/availability-processor/typescript/src/handler.ts
import type { CaregiverPermanentUnavailabilityEvent } from "./events";
import type { VisitRepository } from "./repositories/visits";

/**
 * EventProcessor handles caregiver availability events
 */
export class EventProcessor {
  constructor(private readonly visitRepo: VisitRepository) {}

  async handleEvent(
    event: CaregiverPermanentUnavailabilityEvent
  ): Promise<void> {
    const futureDate = new Date(
      event.effectiveFrom.getTime() + 365 * 24 * 60 * 60 * 1000
    ); // 1 year in the future

    const visits = await this.visitRepo.getCalendar(
      event.caregiverId,
      event.effectiveFrom,
      futureDate
    );

    // Unassign all visits that occur after the permanent unavailability starts
    for (const visit of visits) {
      if (visit.startTime > event.effectiveFrom) {
        await this.visitRepo.unassign(visit.id, event.caregiverId);
      }
    }
  }
}
